src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  SavedObjectsClientWrapperFactory,
+  SavedObjectsClientWrapperOptions,
+  SavedObjectsCreateOptions,
+  SavedObjectsErrorHelpers,
+} from '../../../core/server';
+import {
+  extractDataSourceNamesInVegaSpec,
+  extractVegaSpecFromAttributes,
+  findDataSourceIdbyName,
+} from './utils';
+import { getDataSourceEnabled } from './services';
+
+export const VEGA_VISUALIZATION_CLIENT_WRAPPER_ID = 'vega-visualization-client-wrapper';
+
+export const vegaVisualizationClientWrapper: SavedObjectsClientWrapperFactory = (
+  wrapperOptions: SavedObjectsClientWrapperOptions
+) => {
+  const createForVega = async <T = unknown>(
```

Comment:
for create, I think there are a few validations we can do here when mds is enabled:
1. there must be reference data source fields and data source name fields in the data object
2. the data source references must exist - using bulk get
3. the data source references names fetched from 2 should match with all data source names

that means when we call this create API, we need to be able to get the data source references

Reply:
1. The only way for a Vega spec to be updated through UI is from the save button. The save button, whether overwriting the existing saved object or creating a new visualization, will call `create` to the saved objects client. Thus, from UI perspective, a saved object cannot have a `data_source_name` field but no associated references. The only way a saved object does not have references is if the Vega spec does not specify a `data_source_name` (local cluster queries only).
2. If a non-existent datasource reference (like a datasource that has been deleted) will be filtered out when updating the `references` field because calling bulk get on the object will return the `id` field and an `error` field instead of the `attributes` field.
```Typescript
(await wrapperOptions.client.bulkGet(existingDataSourceReferences)).saved_objects.forEach(
  (object) => {
    // @ts-expect-error
    if (!!object.attributes && !!object.attributes.title) {
      // @ts-expect-error
      existingDataSourceIdToNameMap.set(object.id, object.attributes.title);
    }
  }
);
```
3. This has also been checked. The datasource reference name must be in the `dataSourceNamesSet` generated from the Vega spec.
```Typescript
const newReferences = options?.references?.filter((reference) => {
  if (reference.type !== 'data-source') {
    return true;
  }
  const dataSourceName = existingDataSourceIdToNameMap.get(reference.id);
  if (dataSourceNamesSet.has(dataSourceName)) {
    dataSourceNamesSet.delete(dataSourceName);
    return true;
  }

  return false;
});
```

**Note:** 
Before saving a Vega visualization, the user must press `update` on the visualization renderer. If a user were to type a nonexistent/outdated `data_source_name`, the renderer should catch the error and return the message to the user before saving. This means that a non-existent datasource name cannot be in the Vega spec (unless the user ignores the error message and saves anyway).

There is a chance that the user will edit the data source names in the data sources page but not the Vega spec and export, which will cause a mismatch in the `data_source_name` and the actual reference. However, this edge case is hard to check because if the visualization has more than one data source reference and the names of these references were all changed, we cannot map the old names -> updated names.